prov/sockets/src/sock_ep_msg.c

```diff
@@ -379,440 +378,196 @@ static int sock_ep_cm_getpeer(struct fid_ep *ep, void *addr, size_t *addrlen)
-{
-	char c = 0;
-	int ret = 0;
-	struct sock_cm_msg_list_entry *list_entry;
-
-	list_entry = calloc(1, sizeof(*list_entry) + len);
-	if (!list_entry)
-		return -FI_ENOMEM;
-
-	list_entry->msg_len = len;
-	memcpy(&list_entry->msg[0], msg, len);
-	memcpy(&list_entry->addr, addr, sizeof(*addr));
-	list_entry->fid = fid;
-	list_entry->eq = eq;
-
-	fastlock_acquire(&cm->lock);
-	dlist_insert_tail(&list_entry->entry, &cm->msg_list);
-	fastlock_release(&cm->lock);
-
-	ret = ofi_write_socket(cm->signal_fds[0], &c, 1);
-	if (ret != 1) {
-		SOCK_LOG_DBG("failed to signal\n");
-		ret = -FI_EIO;
-	} else {
-		ret = 0;
-		SOCK_LOG_DBG("Enqueued CM Msg\n");
-	}
-	return ret;
-}
-
-static int sock_ep_cm_send_msg(struct sock_cm_entry *cm,
-			       const struct sockaddr_in *addr,
-			       void *msg, size_t len)
-{
-	int ret;
-#if ENABLE_DEBUG
-	char sa_ip[INET_ADDRSTRLEN] = {0};
-	memcpy(sa_ip, inet_ntoa(addr->sin_addr), INET_ADDRSTRLEN);
-#endif
-	ret = sendto(cm->sock, (char *) msg, len, 0,
-		     (struct sockaddr *) addr, sizeof(*addr));
-	SOCK_LOG_DBG("Sending %d to %s:%d\n", ret, sa_ip, ntohs(addr->sin_port));
-	return (ret == len) ? 0 : -1;
-}
-
-static void sock_ep_cm_release_entry(struct sock_cm_msg_list_entry *msg_entry)
-{
-	struct sock_conn_hdr *msg_hdr;
-	struct fi_eq_cm_entry cm_entry;
-	struct sock_ep *sock_ep;
-
-	msg_hdr = (struct sock_conn_hdr *) msg_entry->msg;
-	if (msg_hdr->type == SOCK_CONN_SHUTDOWN) {
-		memset(&cm_entry, 0, sizeof(cm_entry));
-		cm_entry.fid = msg_entry->fid;
-		sock_ep = container_of(cm_entry.fid, struct sock_ep, ep.fid);
-		if (!sock_ep->attr->cm.shutdown_received) {
-			sock_ep->attr->cm.shutdown_received = 1;
-			sock_ep_disable(&sock_ep->ep);
-			if (sock_eq_report_event(msg_entry->eq, FI_SHUTDOWN, &cm_entry,
-			     sizeof(cm_entry), 0))
-				SOCK_LOG_ERROR("Error in writing to EQ\n");
+	int ret, done = 0;
+	while (done != len) {
+		ret = ofi_write_socket(fd, buf + done, len - done);
+		if (ret < 0) {
+			if (errno == EAGAIN || errno == EWOULDBLOCK)
+				continue;
+			return -FI_EIO;
 		}
-	} else {
-		if (sock_eq_report_error(msg_entry->eq, msg_entry->fid, NULL,
-					 0, FI_ETIMEDOUT, -FI_ETIMEDOUT, NULL, 0))
-			SOCK_LOG_ERROR("failed to report error\n");
+		done += ret;
 	}
-
-	dlist_remove(&msg_entry->entry);
-	free(msg_entry);
+	return 0;
 }
 
-static void sock_ep_cm_flush_msg(struct sock_cm_entry *cm)
+static inline int sock_cm_recv(int fd, void *buf, int len)
 {
-	struct dlist_entry *entry, *next_entry;
-	struct sock_cm_msg_list_entry *msg_entry;
-	fastlock_acquire(&cm->lock);
-	for (entry = cm->msg_list.next; entry != &cm->msg_list;) {
-		msg_entry = container_of(entry,
-					 struct sock_cm_msg_list_entry, entry);
-		next_entry = entry->next;
-
-		if (msg_entry->timestamp_ms != 0 &&
-		    fi_gettime_ms() - msg_entry->timestamp_ms < SOCK_CM_COMM_TIMEOUT) {
-			entry = next_entry;
+	int ret, done = 0;
+	while (done != len) {
+		ret = recv(fd, buf + done, len - done, 0);
```

Comment:
the same as above, failed to compile on windows. could you change it to
`ret = recv(fd, (char*)buf + done, len - done, 0);`
thank you